File: internal/broker/diagnostics.go
package broker

import (
	"errors"
	"fmt"
	"regexp"
	"strings"
	"time"
	"unicode"

	"github.com/segfault-stack/media-cookie-broker/internal/providers"
)

const maxDiagnosticEvents = 1000
const maxDiagnosticDepth = 4

var diagnosticCredentialText = regexp.MustCompile(`(?i)(authorization\s*[:=]\s*\S+|cookie\s*:\s*[^;\s=]+=[^;\s]+|BEGIN [A-Z ]*PRIVATE KEY|\b(bearer|basic)\s+[A-Za-z0-9+/._~=-]{8,}|\b(token|access_token|api_key|apikey|secret|password|passwd|master_key|private_key)\s*=\s*[^\s&,;]+)`)

func ValidateDiagnostics(provider string, batch *DiagnosticBatch) error {
	if batch.Provider != provider {
		return errors.New("provider mismatch")
	}
	if !providers.ValidID(provider) {
		return errors.New("unknown provider")
	}
	if batch.Profile == "" {
		batch.Profile = DefaultProfile
	}
	if !ValidProfileID(batch.Profile) {
		return errors.New("invalid profile")
	}
	if batch.SchemaVersion != DiagnosticsSchemaVersion {
		return fmt.Errorf("schema_version must be %d", DiagnosticsSchemaVersion)
	}
	if len(batch.InstallationID) < 8 || len(batch.InstallationID) > 128 || invalidField(batch.InstallationID) {
		return errors.New("invalid installation_id")
	}
	if len(batch.Events) == 0 || len(batch.Events) > maxDiagnosticEvents {
		return errors.New("invalid event count")
	}
	now := time.Now().UTC()
	for i, event := range batch.Events {
		if event.Timestamp.IsZero() || event.Timestamp.After(now.Add(5*time.Minute)) || event.Timestamp.Before(now.Add(-31*24*time.Hour)) {
			return fmt.Errorf("event %d has invalid timestamp", i)
		}
		if len(event.Type) == 0 || len(event.Type) > 100 || invalidField(event.Type) {
			return fmt.Errorf("event %d has invalid type", i)
		}
		if event.Severity != "info" && event.Severity != "warning" && event.Severity != "error" {
			return fmt.Errorf("event %d has invalid severity", i)
		}
		if event.Details == nil {
			event.Details = map[string]any{}
		}
		if eventProvider, present := event.Details["provider"]; present && eventProvider != provider {
			return fmt.Errorf("event %d has mismatched provider", i)
		}
		if eventProfile, present := event.Details["profile"]; present && eventProfile != batch.Profile {
			return fmt.Errorf("event %d has mismatched profile", i)
		}
		if err := validateDetails(event.Details); err != nil {
			return fmt.Errorf("event %d: %w", i, err)
		}
	}
	return nil
}

func validateDetails(value map[string]any) error {
	return validateDiagnosticObject(value, 0)
}

func validateDiagnosticObject(value map[string]any, depth int) error {
	if depth > maxDiagnosticDepth {
		return errors.New("diagnostic value too deeply nested")
	}
	for key, item := range value {
		if len(key) > 100 || invalidField(key) || forbiddenDiagnosticKey(key) {
			return errors.New("forbidden diagnostic field")
		}
		if err := validateDiagnosticValue(item, depth+1); err != nil {
			return err
		}
	}
	return nil
}
func validateDiagnosticValue(value any, depth int) error {
	if depth > maxDiagnosticDepth {
		return errors.New("diagnostic value too deeply nested")
	}
	switch item := value.(type) {
	case string:
		if len(item) > 4096 || diagnosticCredentialText.MatchString(item) {
			return errors.New("invalid diagnostic text")
		}
	case float64, bool, nil:
		return nil
	case []any:
		if len(item) > 500 {
			return errors.New("diagnostic array too large")
		}
		for _, child := range item {
			if err := validateDiagnosticValue(child, depth+1); err != nil {
				return err
			}
		}
	case map[string]any:
		if err := validateDiagnosticObject(item, depth); err != nil {
			return err
		}
	default:
		return errors.New("invalid diagnostic value")
	}
	return nil
}

func forbiddenDiagnosticKey(key string) bool {
	normalized := normalizeDiagnosticKey(key)
	compact := strings.ReplaceAll(normalized, "_", "")
	parts := strings.FieldsFunc(normalized, func(r rune) bool { return r == '_' })
	for _, part := range parts {
		if part == "password" || part == "passwd" || part == "authorization" || part == "secret" {
			return true
		}
	}
	for i := 0; i+1 < len(parts); i++ {
		pair := parts[i] + "_" + parts[i+1]
		switch pair {
		case "access_token", "refresh_token", "bearer_token", "id_token", "api_key", "auth_header",
			"basic_auth", "bearer_auth", "basic_credentials", "bearer_credentials", "cookie_value",
			"cookie_header", "master_key", "private_key":
			return true
		}
	}
	if normalized == "token" || strings.HasSuffix(normalized, "_token") ||
		normalized == "secret" || strings.HasSuffix(normalized, "_secret") ||
		normalized == "api_key" || normalized == "apikey" || strings.HasSuffix(normalized, "_api_key") || strings.HasSuffix(normalized, "_apikey") ||
		normalized == "auth_header" || strings.HasSuffix(normalized, "_auth_header") ||
		normalized == "basic_auth" || strings.HasSuffix(normalized, "_basic_auth") ||
		normalized == "bearer_auth" || strings.HasSuffix(normalized, "_bearer_auth") ||
		normalized == "basic_credentials" || strings.HasSuffix(normalized, "_basic_credentials") ||
		normalized == "bearer_credentials" || strings.HasSuffix(normalized, "_bearer_credentials") ||
		normalized == "cookie" || normalized == "cookies" || strings.HasSuffix(normalized, "_cookie") || strings.HasSuffix(normalized, "_cookies") ||
		normalized == "cookie_value" || strings.HasSuffix(normalized, "_cookie_value") ||
		normalized == "cookie_header" || strings.HasSuffix(normalized, "_cookie_header") ||
		normalized == "master_key" || strings.HasSuffix(normalized, "_master_key") ||
		normalized == "private_key" || strings.HasSuffix(normalized, "_private_key") ||
		strings.Contains(compact, "apikey") || strings.Contains(compact, "cookievalue") ||
		strings.Contains(compact, "cookieheader") || strings.Contains(compact, "masterkey") ||
		strings.Contains(compact, "privatekey") || strings.Contains(compact, "accesstoken") ||
		strings.Contains(compact, "refreshtoken") || strings.Contains(compact, "bearertoken") {
		return true
	}
	return false
}

func normalizeDiagnosticKey(key string) string {
	var out strings.Builder
	for i, r := range key {
		if unicode.IsUpper(r) && i > 0 {
			out.WriteByte('_')
		}
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			out.WriteRune(unicode.ToLower(r))
		} else if out.Len() > 0 {
			out.WriteByte('_')
		}
	}
	return strings.Trim(out.String(), "_")
}

File: extension/diagnostics.js
import {parseBrokerEndpoint} from "./config.js";
import {getProvider} from "./providers/registry.js";

const BATCHES_KEY = "diagnosticBatches";
const STATE_KEY = "diagnosticState";
const INSTALLATION_KEY = "diagnosticInstallationId";
const RETENTION_MS = 30 * 24 * 60 * 60 * 1000;
const MAX_EVENTS = 10000;
const EVENTS_PER_BATCH = 100;
const MAX_DIAGNOSTIC_DEPTH = 4;
let lock = Promise.resolve();

function sensitiveDiagnosticKey(key) {
  const normalized = String(key).replace(/([a-z0-9])([A-Z])/g, "$1_$2").toLowerCase().replace(/[^a-z0-9]+/g, "_").replace(/^_|_$/g, "");
  const compact = normalized.replaceAll("_", "");
  const parts = normalized.split("_");
  if (parts.some(part => ["password", "passwd", "authorization", "secret"].includes(part))) return true;
  const pairs = new Set(parts.slice(0, -1).map((part, index) => `${part}_${parts[index + 1]}`));
  if (["access_token", "refresh_token", "bearer_token", "id_token", "api_key", "auth_header", "basic_auth", "bearer_auth", "basic_credentials", "bearer_credentials", "cookie_value", "cookie_header", "master_key", "private_key"].some(pair => pairs.has(pair))) return true;
  return normalized === "token" || normalized.endsWith("_token") ||
    normalized === "secret" || normalized.endsWith("_secret") ||
    normalized === "api_key" || normalized === "apikey" || normalized.endsWith("_api_key") || normalized.endsWith("_apikey") ||
    normalized === "auth_header" || normalized.endsWith("_auth_header") ||
    normalized === "basic_auth" || normalized.endsWith("_basic_auth") ||
    normalized === "bearer_auth" || normalized.endsWith("_bearer_auth") ||
    normalized === "basic_credentials" || normalized.endsWith("_basic_credentials") ||
    normalized === "bearer_credentials" || normalized.endsWith("_bearer_credentials") ||
    normalized === "cookie" || normalized === "cookies" || normalized.endsWith("_cookie") || normalized.endsWith("_cookies") ||
    normalized === "cookie_value" || normalized.endsWith("_cookie_value") ||
    normalized === "cookie_header" || normalized.endsWith("_cookie_header") ||
    normalized === "master_key" || normalized.endsWith("_master_key") ||
    normalized === "private_key" || normalized.endsWith("_private_key") ||
    compact.includes("apikey") || compact.includes("cookievalue") || compact.includes("cookieheader") ||
    compact.includes("masterkey") || compact.includes("privatekey") || compact.includes("accesstoken") ||
    compact.includes("refreshtoken") || compact.includes("bearertoken");
}

function redactText(value) {
  return value
    .replace(/(authorization\s*[:=]\s*)[^,;\r\n]+/ig, "$1[redacted]")
    .replace(/(cookie\s*:\s*)[^,\r\n]+/ig, "$1[redacted]")
    .replace(/-----BEGIN [A-Z ]*PRIVATE KEY-----[\s\S]*?-----END [A-Z ]*PRIVATE KEY-----/ig, "[redacted private key]")
    .replace(/\b(bearer|basic)\s+[A-Za-z0-9+/._~=-]+/ig, "$1 [redacted]")
    .replace(/\b(token|access_token|api_key|apikey|secret|password|passwd|master_key|private_key)\s*=\s*[^\s&,;]+/ig, "$1=[redacted]")
    .slice(0, 4096);
}

export function redactDiagnosticValue(value, key = "", depth = 0) {
  if (sensitiveDiagnosticKey(key)) return "[redacted]";
  if (typeof value === "string") return redactText(value);
  if (depth >= MAX_DIAGNOSTIC_DEPTH && value && typeof value === "object") return "[truncated]";
  if (Array.isArray(value)) return value.map(item => redactDiagnosticValue(item, "", depth + 1));
  if (value && typeof value === "object") return Object.fromEntries(Object.entries(value).map(([name, item]) => [name, redactDiagnosticValue(item, name, depth + 1)]));
  return value;
}

async function gzip(value) {
  const stream = new Blob([JSON.stringify(value)]).stream().pipeThrough(new CompressionStream("gzip"));
  const bytes = new Uint8Array(await new Response(stream).arrayBuffer());
  return btoa(String.fromCharCode(...bytes));
}
async function gunzip(value) {
  const bytes = Uint8Array.from(atob(value), char => char.charCodeAt(0));
  return new Response(new Blob([bytes]).stream().pipeThrough(new DecompressionStream("gzip"))).json();
}
async function batches() { return (await chrome.storage.local.get(BATCHES_KEY))[BATCHES_KEY] || []; }

export async function installationId() {
  const current = (await chrome.storage.local.get(INSTALLATION_KEY))[INSTALLATION_KEY];
  if (current) return current;
  const id = crypto.randomUUID();
  await chrome.storage.local.set({[INSTALLATION_KEY]: id});
  return id;
}

export async function logDiagnostic(type, details = {}, severity = "info") {
  lock = lock.then(async () => {
    const event = {timestamp: new Date().toISOString(), type, severity, details: redactDiagnosticValue(details)};
    const stored = await batches();
    let batch = stored.at(-1), events = [];
    if (batch && batch.count < EVENTS_PER_BATCH) events = await gunzip(batch.gzip);
    else { batch = {firstAt: event.timestamp, lastAt: event.timestamp, count: 0, gzip: "", pending: true}; stored.push(batch); }
    events.push(event); batch.lastAt = event.timestamp; batch.count = events.length; batch.gzip = await gzip(events); batch.pending = true;
    let count = stored.reduce((n, item) => n + item.count, 0), cutoff = Date.now() - RETENTION_MS;
    // FIFO pruning: remove the oldest events, not the newest events or browser cookies.
    while (stored.length && (Date.parse(stored[0].lastAt) < cutoff || count > MAX_EVENTS)) {
      const oldest = stored[0];
      const oldestEvents = await gunzip(oldest.gzip);
      const removeForAge = oldestEvents.filter(item => Date.parse(item.timestamp) >= cutoff);
      const keepCount = Math.max(0, Math.min(removeForAge.length, MAX_EVENTS - (count - oldestEvents.length)));
      const keep = removeForAge.slice(Math.max(0, removeForAge.length - keepCount));
      count -= oldestEvents.length;
      if (keep.length) {
        oldest.firstAt = keep[0].timestamp;
        oldest.lastAt = keep.at(-1).timestamp;
        oldest.count = keep.length;
        oldest.gzip = await gzip(keep);
        count += keep.length;
        break;
      }
      stored.shift();
    }
    await chrome.storage.local.set({[BATCHES_KEY]: stored});
  }).catch(() => {});
  return lock;
}

export async function recentDiagnostics() {
  const events = [];
  for (const batch of await batches()) try { events.push(...await gunzip(batch.gzip)); } catch (_) {}
  return events.sort((a, b) => b.timestamp.localeCompare(a.timestamp));
}
export async function diagnosticsState() {
  const stored = await batches(), state = (await chrome.storage.local.get(STATE_KEY))[STATE_KEY] || {};
  const pending = stored.filter(item => item.pending);
  return {...state, queuedBatches: pending.length, queuedEvents: pending.reduce((n, item) => n + item.count, 0)};
}
export async function clearDiagnostics() { await chrome.storage.local.set({[BATCHES_KEY]: [], [STATE_KEY]: {lastClearedAt: new Date().toISOString()}}); }

export async function uploadDiagnostics(cfg, basic) {
  if (cfg.remoteDiagnosticsEnabled !== true) return true;
  const state = (await chrome.storage.local.get(STATE_KEY))[STATE_KEY] || {};
  if (!cfg.password || state.nextAttemptAt > Date.now()) return false;
  const stored = await batches(), pending = stored.filter(item => item.pending).slice(0, 10);
  if (!pending.length) return true;
  const selected = [];
  const grouped = new Map();
  const completed = new Set();
  const updateBatches = () => stored.map(batch => {
    const item = selected.find(candidate => candidate.batch === batch);
    if (!item) return batch;
    const uploadedScopes = new Set(batch.uploadedScopes || []);
    for (const scope of completed) uploadedScopes.add(scope);
    const stillPending = [...item.neededScopes].some(scope => !uploadedScopes.has(scope));
    return {...batch, uploadedScopes: [...uploadedScopes], pending: stillPending};
  });
  try {
    for (const batch of pending) {
      const events = await gunzip(batch.gzip);
      const neededScopes = new Set();
      const uploadedScopes = new Set(batch.uploadedScopes || []);
      for (const event of events) {
        const provider = event.details?.provider;
        if (!getProvider(provider) || !cfg.enabledProviders.includes(provider)) continue;
        const profile = event.details?.profile || "default";
        const scope = `${provider}/${profile}`;
        neededScopes.add(scope);
        if (uploadedScopes.has(scope)) continue;
        if (!grouped.has(scope)) grouped.set(scope, {provider, profile, events: []});
        grouped.get(scope).events.push(event);
      }
      selected.push({batch, neededScopes});
    }
    for (const [scope, group] of grouped) {
      const data = JSON.stringify({schema_version: 1, provider: group.provider, profile: group.profile, installation_id: await installationId(), events: group.events});
      const compressed = await new Response(new Blob([data]).stream().pipeThrough(new CompressionStream("gzip"))).arrayBuffer();
      const url = new URL(parseBrokerEndpoint(cfg.endpoint));
      url.pathname = `${url.pathname.replace(/\/$/, "")}/v1/diagnostics/events`;
      const response = await fetch(url, {
        method: "POST",
        headers: {Authorization: basic(cfg.username, cfg.password), "Content-Type": "application/json", "Content-Encoding": "gzip"},
        body: compressed
      });
      if (!response.ok) throw new Error(`${scope}: HTTP ${response.status}`);
      completed.add(scope);
    }
    await chrome.storage.local.set({
      [BATCHES_KEY]: updateBatches(),
      [STATE_KEY]: {lastUploadedAt: new Date().toISOString(), failures: 0}
    });
    return true; // Successful acknowledgements are deliberately not logged.
  } catch (error) {
    const failures = Number(state.failures || 0) + 1;
    await chrome.storage.local.set({
      [BATCHES_KEY]: updateBatches(),
      [STATE_KEY]: {failures, lastError: String(error?.message || error), nextAttemptAt: Date.now() + Math.min(3600000, 1000 * 2 ** failures)}
    });
    return false;
  }
}
